### DISCUSSION TITLE
Inject services into activities retrieved from calling service scope

### DISCUSSION THREAD

**Discussion by JochenSteinbrenner** [elsa-workflows/elsa-core#4203] [Q&A]
Hello to all,
at the moment our program is based on ELSA V1 and its working very well.
We consider to migrate to ELSA V3.
One necessary point for is, that we have to inject custom scoped services into our custom activities.
Within V1, the services can be injected into the activities and they are retrieved from
the service scope which calls the workflow.
Means the scoped services within activities are exactly the same as within the calling scope.

Within V3, ELSA creates a new scope within workflow triggers. Means, the services, injected to the activity
are not the same services as within the calling scope.

Within following example, is it possible to inject the created entitySystem instance
into the findEntity activity?

The reason is, that we call many workflows within a scope which all can change our business data. On exceptions, we want to rollback
all changes of all workflows called in the scope. Therefore we have implemented a transaction mechanism within the entitySystem instance in order to rollback all changes.

Thank you for help! Great work!
with kind regards

Jochen Steinbrenner
AZO

```
using (var serviceScope = serviceScopeFactory.CreateScope())
{
    var entitySystem = serviceScope.ServiceProvider.GetService<IEntitySystem>();
    var options = new StartWorkflowRuntimeOptions()
    {
        VersionOptions = VersionOptions.Published,
        Input = input
    };
    var workflowExecutionResult = await workflowRuntime.StartWorkflowAsync(workflowDefinition.DefinitionId, options);
    return workflowExecutionResult;
}
```

```
public class FindEntityActivity : CodeActivity
    {
        [JsonConstructor]
        public FindEntityActivity()
        {
        }
        protected override void Execute(ActivityExecutionContext context)
        {
            var entitySystem = context.GetService<IEntitySystem>();
            var s = context.Get(EntityGuid) ?? Guid.Empty.ToString();
            var entityGuid = Guid.Parse(s);
            var entity = entitySystem.FindEntity(entityGuid);
            Variable.Set(context, entity);
        }

        [Input(Description = "The variable to assign the value to.")]
        public Variable Variable { get; set; } = default!;
       
        [Input(Description = "The entity guid.")]
        public Input<string?> EntityGuid { get; set; } = new("00000000-0000-0000-0000-000000000000");
    }
```

**Comment by JochenSteinbrenner** [comment]
Hello, is there any possible solution for my problem?

**Comment by sfmskywalker** [comment]
I've been thinking about this, and I came up with 2 options we could consider trying out:

1. We could consider adding an option to the `IWorkflowRuntime` service to pass in application context, which would be available to all activities. This application context would be a dictionary of objects. Your custom activity would have knowledge of this dictionary and where to find the entity system. A disadvantage of this approach might be tight-coupling of your activity with calling code that **must** provide this application context.

2. Another option might be to provide access to the service provider that instantiated the workflow runtime. You would have to change your invoking code a little bit though, by resolving the workflow runtime from the same scope as the entity system. Example:

```csharp
using (var serviceScope = serviceScopeFactory.CreateScope())
{
    var entitySystem = serviceScope.ServiceProvider.GetService<IEntitySystem>();
    var options = new StartWorkflowRuntimeOptions()
    {
        VersionOptions = VersionOptions.Published,
        Input = input
    };

    // Resolve workflow runtime from same scope.
    var workflowRuntime = serviceScope.ServiceProvider.GetRequiredService<IWorkflowRuntime>();
    var workflowExecutionResult = await workflowRuntime.StartWorkflowAsync(workflowDefinition.DefinitionId, options);
    return workflowExecutionResult;
}
```

Then your activity code would be able to do this:

```csharp
public class FindEntityActivity : CodeActivity
    {
        [JsonConstructor]
        public FindEntityActivity()
        {
        }
        protected override void Execute(ActivityExecutionContext context)
        {
            var serviceProvider = context.RootServiceProvider; // This would be the same provider from which IWorkflowRuntimeService was resolved, therefore also providing access to the same scoped entity system.
            var entitySystem = serviceProvider.GetService<IEntitySystem>();
        }
    }
```

In either option we would have to make some adjustments to Elsa 3. I think the second option is the cleanest, but I'm open for alternative suggestions.

**Reply by JochenSteinbrenner** [reply]
Hello Sipke, first of all, thank you for your answer!
I prefer the second solution in general. The implementation
is very easy for me and it works for many customer scoped services
and many custom activities without much effort. Sounds good.

Can you please consider to add this?
Many thanks!

regards
Jochen Steinbrenner

**Reply by JochenSteinbrenner** [reply]
Good morning, one thing i've tried out:
- Create and register a custom AzoWorkflowHostFactory
- CreateAsync produces a custom AzoWorkflowHost
- AzoWorkflowHost.Start/Resume
  - Create workflowExecutionContext with the scoped serviceProvider
  - AzoWorkflowHost calls WorkflowRunner.RunAsync(WorkflowExecutionContext)
 
At the moment I struggle with getting the scoped service provider into the AzoWorkflowHost.
Maybe this aprouch can be an alternative for you...

**Reply by sfmskywalker** [reply]
Hi Jochen,

I created #4314 to track this change request for the 2nd option. I hope that in the meantime you were able to work around this limitation.
